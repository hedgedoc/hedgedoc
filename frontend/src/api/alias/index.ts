/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { DeleteApiRequestBuilder } from '../common/api-request-builder/delete-api-request-builder'
import { PostApiRequestBuilder } from '../common/api-request-builder/post-api-request-builder'
import { PutApiRequestBuilder } from '../common/api-request-builder/put-api-request-builder'
import type { Alias, NewAliasDto, PrimaryAliasDto } from './types'

/**
 * Adds an alias to an existing note.
 *
 * @param noteIdOrAlias The note id or an existing alias for a note.
 * @param newAlias The new alias.
 * @return Information about the newly created alias.
 * @throws {Error} when the api request wasn't successful
 */
export const addAlias = async (noteIdOrAlias: string, newAlias: string): Promise<Alias> => {
  const response = await new PostApiRequestBuilder<Alias, NewAliasDto>('alias')
    .withJsonBody({
      noteIdOrAlias,
      newAlias
    })
    .sendRequest()
  return response.asParsedJsonObject()
}

/**
 * Marks a given alias as the primary one for a note.
 * The former primary alias should be marked as non-primary by the backend automatically.
 *
 * @param alias The alias to mark as primary for its corresponding note.
 * @return The updated information about the alias.
 * @throws {Error} when the api request wasn't successfull
 */
export const markAliasAsPrimary = async (alias: string): Promise<Alias> => {
  const response = await new PutApiRequestBuilder<Alias, PrimaryAliasDto>('alias/' + alias)
    .withJsonBody({
      primaryAlias: true
    })
    .sendRequest()
  return response.asParsedJsonObject()
}

/**
 * Removes a given alias from its corresponding note.
 *
 * @param alias The alias to remove from its note.
 * @throws {Error} when the api request wasn't successful.
 */
export const deleteAlias = async (alias: string): Promise<void> => {
  await new DeleteApiRequestBuilder('alias/' + alias).sendRequest()
}
