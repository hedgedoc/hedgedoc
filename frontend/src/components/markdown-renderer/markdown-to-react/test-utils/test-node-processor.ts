/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { NodeProcessor } from '../../node-preprocessors/node-processor'
import type { Document } from 'domhandler'
import { Element, Text } from 'domhandler'

export class TestNodeProcessor extends NodeProcessor {
  process(document: Document): Document {
    document.childNodes.push(new Element('node-processor', {}, [new Text('node processor children')]))
    return document
  }
}
