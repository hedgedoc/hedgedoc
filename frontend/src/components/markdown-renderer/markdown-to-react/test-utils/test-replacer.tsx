/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { NativeRenderer, NodeReplacement, SubNodeTransform } from '../../replace-components/component-replacer'
import { ComponentReplacer, DO_NOT_REPLACE } from '../../replace-components/component-replacer'
import type { Element } from 'domhandler'
import React, { Fragment } from 'react'

export class TestReplacer extends ComponentReplacer {
  replace(node: Element, subNodeTransform: SubNodeTransform, nativeRenderer: NativeRenderer): NodeReplacement {
    return node.tagName === 'node-processor' ? (
      <Fragment>
        <span>NodeProcessor! </span>
        <span data-children={true}> {node.childNodes.map(subNodeTransform)} </span>
        <span data-native={true}> {nativeRenderer()} </span>
      </Fragment>
    ) : (
      DO_NOT_REPLACE
    )
  }
}
