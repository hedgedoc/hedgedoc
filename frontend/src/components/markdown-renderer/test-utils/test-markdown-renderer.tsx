/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { StoreProvider } from '../../../redux/store-provider'
import type { MarkdownRendererExtension } from '../extensions/_base-classes/markdown-renderer-extension'
import { MarkdownToReact } from '../markdown-to-react/markdown-to-react'
import React, { useMemo } from 'react'

export interface SimpleMarkdownRendererProps {
  content: string
  extensions: MarkdownRendererExtension[]
}

/**
 * A markdown renderer for tests.
 *
 * @param content The content to be rendered.
 * @param extensions The {@link MarkdownRendererExtension MarkdownExtensions} to use for rendering.
 */
export const TestMarkdownRenderer: React.FC<SimpleMarkdownRendererProps> = ({ content, extensions }) => {
  const lines = useMemo(() => content.split('\n'), [content])
  return (
    <StoreProvider>
      <MarkdownToReact
        markdownContentLines={lines}
        markdownRenderExtensions={extensions}
        newlinesAreBreaks={true}
        allowHtml={false}
      />
    </StoreProvider>
  )
}
