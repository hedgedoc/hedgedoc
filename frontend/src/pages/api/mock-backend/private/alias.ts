/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { Alias, NewAliasDto } from '../../../../api/alias/types'
import { HttpMethod, respondToMatchingRequest } from '../../../../handler-utils/respond-to-matching-request'
import type { NextApiRequest, NextApiResponse } from 'next'

const handler = (req: NextApiRequest, res: NextApiResponse) => {
  respondToMatchingRequest<Alias>(
    HttpMethod.POST,
    req,
    res,
    {
      name: (req.body as NewAliasDto).newAlias,
      noteId: (req.body as NewAliasDto).noteIdOrAlias,
      primaryAlias: false
    },
    201
  )
}

export default handler
