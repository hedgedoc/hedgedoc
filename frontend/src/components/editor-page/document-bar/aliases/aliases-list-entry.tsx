/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { deleteAlias, markAliasAsPrimary } from '../../../../api/alias'
import type { Alias } from '../../../../api/alias/types'
import { updateMetadata } from '../../../../redux/note-details/methods'
import { testId } from '../../../../utils/test-id'
import { ForkAwesomeIcon } from '../../../common/fork-awesome/fork-awesome-icon'
import { ShowIf } from '../../../common/show-if/show-if'
import { useUiNotifications } from '../../../notifications/ui-notification-boundary'
import React, { useCallback } from 'react'
import { Button } from 'react-bootstrap'
import { useTranslation } from 'react-i18next'

export interface AliasesListEntryProps {
  alias: Alias
}

/**
 * Component that shows an entry in the aliases list with buttons to remove it or mark it as primary.
 *
 * @param alias The alias.
 */
export const AliasesListEntry: React.FC<AliasesListEntryProps> = ({ alias }) => {
  const { t } = useTranslation()
  const { showErrorNotification } = useUiNotifications()

  const onRemoveClick = useCallback(() => {
    deleteAlias(alias.name)
      .then(updateMetadata)
      .catch(showErrorNotification(t('editor.modal.aliases.errorRemovingAlias')))
  }, [alias, t, showErrorNotification])

  const onMakePrimaryClick = useCallback(() => {
    markAliasAsPrimary(alias.name)
      .then(updateMetadata)
      .catch(showErrorNotification(t('editor.modal.aliases.errorMakingPrimary')))
  }, [alias, t, showErrorNotification])

  return (
    <li className={'list-group-item d-flex flex-row justify-content-between align-items-center'}>
      {alias.name}
      <div>
        <ShowIf condition={alias.primaryAlias}>
          <Button
            className={'me-2 text-warning'}
            variant='light'
            disabled={true}
            title={t('editor.modal.aliases.isPrimary') ?? undefined}
            {...testId('aliasIsPrimary')}>
            <ForkAwesomeIcon icon={'star'} />
          </Button>
        </ShowIf>
        <ShowIf condition={!alias.primaryAlias}>
          <Button
            className={'me-2'}
            variant='light'
            title={t('editor.modal.aliases.makePrimary') ?? undefined}
            onClick={onMakePrimaryClick}
            {...testId('aliasButtonMakePrimary')}>
            <ForkAwesomeIcon icon={'star-o'} />
          </Button>
        </ShowIf>
        <Button
          variant='light'
          className={'text-danger'}
          title={t('editor.modal.aliases.removeAlias') ?? undefined}
          onClick={onRemoveClick}
          {...testId('aliasButtonRemove')}>
          <ForkAwesomeIcon icon={'times'} />
        </Button>
      </div>
    </li>
  )
}
