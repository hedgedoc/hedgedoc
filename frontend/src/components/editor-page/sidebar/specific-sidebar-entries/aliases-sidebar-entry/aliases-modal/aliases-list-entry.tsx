/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { deleteAlias, markAliasAsPrimary } from '../../../../../../api/alias'
import type { Alias } from '../../../../../../api/alias/types'
import { useIsOwner } from '../../../../../../hooks/common/use-is-owner'
import { useTranslatedText } from '../../../../../../hooks/common/use-translated-text'
import { updateMetadata } from '../../../../../../redux/note-details/methods'
import { testId } from '../../../../../../utils/test-id'
import { UiIcon } from '../../../../../common/icons/ui-icon'
import { useUiNotifications } from '../../../../../notifications/ui-notification-boundary'
import React, { useCallback } from 'react'
import { Badge } from 'react-bootstrap'
import { Button } from 'react-bootstrap'
import { Star as IconStar, X as IconX } from 'react-bootstrap-icons'
import { useTranslation, Trans } from 'react-i18next'

export interface AliasesListEntryProps {
  alias: Alias
}

/**
 * Component that shows an entry in the aliases list with buttons to remove it or mark it as primary.
 *
 * @param alias The alias.
 */
export const AliasesListEntry: React.FC<AliasesListEntryProps> = ({ alias }) => {
  const { t } = useTranslation()
  const { showErrorNotification } = useUiNotifications()
  const isOwner = useIsOwner()

  const onRemoveClick = useCallback(() => {
    deleteAlias(alias.name)
      .then(updateMetadata)
      .catch(showErrorNotification(t('editor.modal.aliases.errorRemovingAlias')))
  }, [alias, t, showErrorNotification])

  const onMakePrimaryClick = useCallback(() => {
    markAliasAsPrimary(alias.name)
      .then(updateMetadata)
      .catch(showErrorNotification(t('editor.modal.aliases.errorMakingPrimary')))
  }, [alias, t, showErrorNotification])

  const isPrimaryText = useTranslatedText('editor.modal.aliases.isPrimary')
  const makePrimaryText = useTranslatedText('editor.modal.aliases.makePrimary')
  const removeAliasText = useTranslatedText('editor.modal.aliases.removeAlias')

  return (
    <li className={'list-group-item d-flex flex-row justify-content-between align-items-center'}>
      <div>
        {alias.name}
        {alias.primaryAlias && (
          <Badge bg='secondary' className={'ms-2'} title={isPrimaryText} {...testId('aliasPrimaryBadge')}>
            <Trans i18nKey={'editor.modal.aliases.primaryLabel'}></Trans>
          </Badge>
        )}
      </div>
      <div>
        {!alias.primaryAlias && (
          <Button
            className={'me-2'}
            variant='secondary'
            disabled={!isOwner}
            title={makePrimaryText}
            onClick={onMakePrimaryClick}
            {...testId('aliasButtonMakePrimary')}>
            <UiIcon icon={IconStar} />
          </Button>
        )}
        <Button
          variant='danger'
          disabled={!isOwner}
          title={removeAliasText}
          onClick={onRemoveClick}
          {...testId('aliasButtonRemove')}>
          <UiIcon icon={IconX} />
        </Button>
      </div>
    </li>
  )
}
