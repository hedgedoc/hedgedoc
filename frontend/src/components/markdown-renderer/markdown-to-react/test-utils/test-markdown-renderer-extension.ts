/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { MarkdownRendererExtension } from '../../extensions/_base-classes/markdown-renderer-extension'
import type { NodeProcessor } from '../../node-preprocessors/node-processor'
import type { ComponentReplacer } from '../../replace-components/component-replacer'
import { TestNodeProcessor } from './test-node-processor'
import { TestReplacer } from './test-replacer'
import type MarkdownIt from 'markdown-it'
import Token from 'markdown-it/lib/token'

export class TestMarkdownRendererExtension extends MarkdownRendererExtension {
  buildNodeProcessors(): NodeProcessor[] {
    return [new TestNodeProcessor()]
  }

  configureMarkdownIt(markdownIt: MarkdownIt) {
    markdownIt.use(() => {
      markdownIt.core.ruler.push('configure', (core) => core.tokens.push(new Token('configure', 'configure', 0)))
      markdownIt.renderer.rules.configure = () => '<span>configure</span>'
    })
  }

  buildReplacers(): ComponentReplacer[] {
    return [new TestReplacer()]
  }

  buildTagNameAllowList(): string[] {
    return ['node-processor']
  }

  configureMarkdownItPost(markdownIt: MarkdownIt) {
    markdownIt.use(() => {
      markdownIt.core.ruler.push('post', (core) => core.tokens.push(new Token('post', 'post', 0)))
      markdownIt.renderer.rules.post = () => '<span>post</span>'
    })
  }
}
