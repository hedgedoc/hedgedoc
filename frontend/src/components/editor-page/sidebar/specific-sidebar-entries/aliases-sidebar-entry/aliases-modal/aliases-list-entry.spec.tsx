/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import * as AliasModule from '../../../../../../api/alias'
import type { Alias } from '../../../../../../api/alias/types'
import * as NoteDetailsReduxModule from '../../../../../../redux/note-details/methods'
import { mockI18n } from '../../../../../../test-utils/mock-i18n'
import { mockNotePermissions } from '../../../../../../test-utils/mock-note-permissions'
import { AliasesListEntry } from './aliases-list-entry'
import { act, render, screen } from '@testing-library/react'
import React from 'react'
import { mockUiNotifications } from '../../../../../../test-utils/mock-ui-notifications'

jest.mock('../../../../../../api/alias')
jest.mock('../../../../../../redux/note-details/methods')
jest.mock('../../../../../notifications/ui-notification-boundary')
jest.mock('../../../../../../hooks/common/use-application-state')

const deletePromise = Promise.resolve()
const markAsPrimaryPromise = Promise.resolve({ name: 'mock', primaryAlias: true, noteId: 'mock' })

describe('AliasesListEntry', () => {
  beforeEach(async () => {
    await mockI18n()
    mockUiNotifications()
    jest.spyOn(AliasModule, 'deleteAlias').mockImplementation(() => deletePromise)
    jest.spyOn(AliasModule, 'markAliasAsPrimary').mockImplementation(() => markAsPrimaryPromise)
    jest.spyOn(NoteDetailsReduxModule, 'updateMetadata').mockImplementation(() => Promise.resolve())
  })

  afterEach(() => {
    jest.resetAllMocks()
    jest.resetModules()
  })

  it('renders an AliasesListEntry that is primary', async () => {
    mockNotePermissions('test', 'test')
    const testAlias: Alias = {
      name: 'test-primary',
      primaryAlias: true,
      noteId: 'test-note-id'
    }
    const view = render(<AliasesListEntry alias={testAlias} />)
    expect(view.container).toMatchSnapshot()
    const button = await screen.findByTestId('aliasButtonRemove')
    await act<void>(() => {
      button.click()
    })
    expect(AliasModule.deleteAlias).toBeCalledWith(testAlias.name)
    await deletePromise
    expect(NoteDetailsReduxModule.updateMetadata).toBeCalled()
  })

  it("adds aliasPrimaryBadge & removes aliasButtonMakePrimary in AliasesListEntry if it's primary", () => {
    mockNotePermissions('test2', 'test')
    const testAlias: Alias = {
      name: 'test-primary',
      primaryAlias: true,
      noteId: 'test-note-id'
    }
    const view = render(<AliasesListEntry alias={testAlias} />)
    expect(view.container).toMatchSnapshot()
  })

  it('renders an AliasesListEntry that is not primary', async () => {
    mockNotePermissions('test', 'test')
    const testAlias: Alias = {
      name: 'test-non-primary',
      primaryAlias: false,
      noteId: 'test-note-id'
    }
    const view = render(<AliasesListEntry alias={testAlias} />)
    expect(view.container).toMatchSnapshot()
    const buttonRemove = await screen.findByTestId('aliasButtonRemove')
    await act<void>(() => {
      buttonRemove.click()
    })
    expect(AliasModule.deleteAlias).toBeCalledWith(testAlias.name)
    await deletePromise
    expect(NoteDetailsReduxModule.updateMetadata).toBeCalled()
    const buttonMakePrimary = await screen.findByTestId('aliasButtonMakePrimary')
    await act<void>(() => {
      buttonMakePrimary.click()
    })
    expect(AliasModule.markAliasAsPrimary).toBeCalledWith(testAlias.name)
    await markAsPrimaryPromise
    expect(NoteDetailsReduxModule.updateMetadata).toBeCalled()
  })

  it("removes aliasPrimaryBadge & adds aliasButtonMakePrimary in AliasesListEntry if it's not primary", () => {
    mockNotePermissions('test2', 'test')
    const testAlias: Alias = {
      name: 'test-primary',
      primaryAlias: false,
      noteId: 'test-note-id'
    }
    const view = render(<AliasesListEntry alias={testAlias} />)
    expect(view.container).toMatchSnapshot()
  })
})
