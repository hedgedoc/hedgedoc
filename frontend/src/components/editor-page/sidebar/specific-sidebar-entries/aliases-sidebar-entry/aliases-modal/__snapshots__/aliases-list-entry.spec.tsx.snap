// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AliasesListEntry adds aliasPrimaryBadge & removes aliasButtonMakePrimary in AliasesListEntry if it's primary 1`] = `
<div>
  <li
    class="list-group-item d-flex flex-row justify-content-between align-items-center"
  >
    <div>
      test-primary
      <span
        class="ms-2 badge bg-secondary"
        data-testid="aliasPrimaryBadge"
        title="editor.modal.aliases.isPrimary"
      >
        editor.modal.aliases.primaryLabel
      </span>
    </div>
    <div>
      <button
        class="btn btn-danger"
        data-testid="aliasButtonRemove"
        disabled=""
        title="editor.modal.aliases.removeAlias"
        type="button"
      >
        BootstrapIconMock_X
      </button>
    </div>
  </li>
</div>
`;

exports[`AliasesListEntry removes aliasPrimaryBadge & adds aliasButtonMakePrimary in AliasesListEntry if it's not primary 1`] = `
<div>
  <li
    class="list-group-item d-flex flex-row justify-content-between align-items-center"
  >
    <div>
      test-primary
    </div>
    <div>
      <button
        class="me-2 btn btn-secondary"
        data-testid="aliasButtonMakePrimary"
        disabled=""
        title="editor.modal.aliases.makePrimary"
        type="button"
      >
        BootstrapIconMock_Star
      </button>
      <button
        class="btn btn-danger"
        data-testid="aliasButtonRemove"
        disabled=""
        title="editor.modal.aliases.removeAlias"
        type="button"
      >
        BootstrapIconMock_X
      </button>
    </div>
  </li>
</div>
`;

exports[`AliasesListEntry renders an AliasesListEntry that is not primary 1`] = `
<div>
  <li
    class="list-group-item d-flex flex-row justify-content-between align-items-center"
  >
    <div>
      test-non-primary
    </div>
    <div>
      <button
        class="me-2 btn btn-secondary"
        data-testid="aliasButtonMakePrimary"
        title="editor.modal.aliases.makePrimary"
        type="button"
      >
        BootstrapIconMock_Star
      </button>
      <button
        class="btn btn-danger"
        data-testid="aliasButtonRemove"
        title="editor.modal.aliases.removeAlias"
        type="button"
      >
        BootstrapIconMock_X
      </button>
    </div>
  </li>
</div>
`;

exports[`AliasesListEntry renders an AliasesListEntry that is primary 1`] = `
<div>
  <li
    class="list-group-item d-flex flex-row justify-content-between align-items-center"
  >
    <div>
      test-primary
      <span
        class="ms-2 badge bg-secondary"
        data-testid="aliasPrimaryBadge"
        title="editor.modal.aliases.isPrimary"
      >
        editor.modal.aliases.primaryLabel
      </span>
    </div>
    <div>
      <button
        class="btn btn-danger"
        data-testid="aliasButtonRemove"
        title="editor.modal.aliases.removeAlias"
        type="button"
      >
        BootstrapIconMock_X
      </button>
    </div>
  </li>
</div>
`;
